### ISSUE TITLE
gopsutil missing a way to retrieve CPU architecture like in 'uname -m'

### ISSUE THREAD

**Issue by tmm1** [shirou/gopsutil#737]
I was surprised to see that the equivalent of `uname -m` is not available anywhere.

**Comment by Lomanic** [comment]
There is `runtime.GOARCH` though it only returns the arch the binary has been compiled for, which can be different from real CPU running it on OSes able to run both 32 and 64 bit binaries

You may look at the [`cpu.Info()`](https://godoc.org/github.com/shirou/gopsutil/cpu#Info) function, though it doesn't result the same thing as `uname -m`. You may check [github.com/klauspost/cpuid](https://github.com/klauspost/cpuid) for CPUs with the CPUID instruction (ARM not supported).

Maybe I overlooked it, but indeed it doesn't seem there is a way to get hardware architecture like `uname -m` (or runtime.GOARCH) in gopsutil. Not sure if it should be obtained in `cpu.Info()` or [`host.Info()`](https://godoc.org/github.com/shirou/gopsutil/host#Info).

**Comment by tmm1** [comment]
The cpu package provides some hints, but I'm more interested in the OS architecture. For instance on aarch64 process its still possible to run kernel/OS in armv8l mode.

I think we should add the equivalent of `uname -m` to `host.Info()`. This may be same as `runtime.GOARCH` in most cases, but not all.

**Comment by Lomanic** [comment]
Implementations in python
https://github.com/pydata/numexpr/blob/master/numexpr/cpuinfo.py
https://github.com/workhorsy/py-cpuinfo/blob/master/cpuinfo/cpuinfo.py

The objective to me is to return the same result as runtime.GOARCH, while really querying at runtime instead of relying on a constant set at build-time.

Possible values (obtained with `go tool dist list | cut -d/ -f2 | sort -u`)

```
386
amd64
amd64p32
arm
arm64
mips
mips64
mips64le
mipsle
ppc64
ppc64le
s390x
wasm
```

**Comment by Lomanic** [comment]
## Notes for implementing

On Unix, use uname syscall like below and compare returned value to [this list](https://stackoverflow.com/a/45125525) to do the mapping with the possible GOARCH values above

```go
package main

import (
	"fmt"

	"golang.org/x/sys/unix"
)

func main() {
	var utsname unix.Utsname
	unix.Uname(&utsname)
	fmt.Println(string(utsname.Machine[:]))
}
```

For Windows, call [GetNativeSystemInfo](https://docs.microsoft.com/fr-fr/windows/win32/api/sysinfoapi/nf-sysinfoapi-getnativesysteminfo) (or GetSystemInfo if not available) and do the mapping between [win_SystemInfo.wProcessorArchitecture](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-system_info) and the possible GOARCH values.

**Comment by tmm1** [comment]
>The objective to me is to return the same result as runtime.GOARCH, while really querying at runtime instead of relying on a constant set at build-time.

This sounds fine, but I think it is also important to return the underlying value. For my use-case I already know GOARCH=arm but I need to know the specific variant of arm i.e. armv8l or armv8b

**Comment by Lomanic** [comment]
I would prefer returned values to be consistent between Unix and Windows, hence the mapping with runtime.GOARCH values, that Go developers are already used to. If you need the exact value of `uname -m`, then I think you should call yourself `unix.Uname()` like in my previous reply instead of relying on gopsutil for this.

In fact you can get this information with `Infostat.ModelName` returned by `cpu.Info()` (returns `ARMv7 Processor rev 1 (v7l)` on an armv7l Linux machine) if you really want to use only gopsutil.

**Comment by tmm1** [comment]
I am already using cpu.Info, but cpu arch and host arch are not the same thing. I have a aarch64 cpu but running armv8l kernel. Also many ARM cpus do not publish and vendor or model info to /proc/cpuinfo

IMHO, along with KernelVersion there should be a KernelArch field which includes the raw arch value from uname or windows equivalent.

**Comment by Lomanic** [comment]
Thanks for the explanation. I started the work on a new KernelArch field of host.InfoStat as returned by host.Info(), getting its content from uname syscall (as-is) on POSIX platforms and GetNativeSystemInfo on Windows (with some mapping done to return the right corresponding "uname-like" values).

I still would like to return a proper value queried at runtime (instead of depending on a build constant) à la runtime.GOOS, I will implement it in the future but it will then be easy to do the mapping as OS-agnostic.

**Comment by shirou** [comment]
merged #745  Thank you for joining the discussion and implementation!
